### PR TITLE
Fix android test coverage

### LINKED ISSUE #16640: Emulator tests not actually registering coverage
### Checked for duplicates?

- [X] This issue is not a duplicate

### Does it also happen in the desktop version?

- [X] This bug does not occur in the latest version of Anki

### What are the steps to reproduce this bug?

- `./gradlew jacocoAndroidTestReport`
- open the coverage report (Anki-Android/AnkiDroid/build/reports/jacoco/jacocoAndroidTestReport/html/index.html)
- check for any coverage reported at all

### Expected behaviour

coverage reported

### Actual behaviour

zero coverage

### Debug info

```text
current main as of this typing (or release-2.18, for that matter)
```

### (Optional) Anything else you want to share?

An automated way to see if any coverage at all was registered *appears* to be to look for the final summary statistic in the XML report generated:

> cat AnkiDroid/build/reports/jacoco/jacocoAndroidTestReport/jacocoAndroidTestReport.xml |grep 'covered="0"/></report>'

It's self evident that there should be *any* number but zero in there

release-2.17 has coverage coming through, release-2.18 does not

### Research

- [X] I am reporting a bug specific to AnkiDroid (Android app)
- [X] I have checked the [manual](https://ankidroid.org/docs/manual.html) and the [FAQ](https://github.com/ankidroid/Anki-Android/wiki/FAQ) and could not find a solution to my issue
- [ ] (Optional) I have confirmed the issue is not resolved in the latest alpha release ([instructions](https://docs.ankidroid.org/manual.html#betaTesting))

### PR DESCRIPTION
**Pull request by mikehardy**
## Purpose / Description

Emulator tests currently run, but they generate zero coverage stats despite seemingly succeeding

Root cause is an incompatibility between android gradle plugin >= 8.3.0 *and* a common gradle lib that is used as a dependency in both testImplementation and androidTestImplementation

tracked upstream in https://issuetracker.google.com/issues/332746900

## Fixes
* Fixes #16640

## Approach

- we don't want to back down android gradle plugin versions, so we need to work around this
- we don't want to actually duplicate common testlib code
- so...use a gradle copy task to copy in the common testlib code at build time (after de-conflicting path)
- ...now we have the common code in androidTest as well and can remove the androidTestImplementation dep
- ...then use a gradle finalization task that verifies coverage actually happened

It's a bit unseemly but it works ?

## How Has This Been Tested?

Lots and lots and lots of `./gradlew jacocoAndroidTestReport` then checking that the file ends with non-zero coverage
A fair bit of `./gradlew copyTestLibIntoAndroidTest` to verify the copy works
A fair bit of `./gradlew assertNonzeroAndroidTestCoverage` to get the regex matching working well

## Learning (optional, can help others)

Most people must not really care about testing because the testing area is full of subtle bugs for cases that seem normal to me, and the tools constantly have erroneous interactions that regress / recur in new ways with new versions

## Checklist
_Please, go through these checks before submitting the PR._

- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

**Comment by mikehardy** [comment]
Okay, need to fix some stuff but it's gradle dependency declaration items, not logical things (those are working)

```
    Reason: Task ':AnkiDroid:runKtlintCheckOverAndroidTestSourceSet' uses this output of task ':AnkiDroid:copyTestLibIntoAndroidTest' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed.
    ```

**Comment by mikehardy** [comment]
If you follow the codecov links it looks like we get a bit more than 4% coverage back (a 10% bump or so...) as coverage is now coming through correctly from what I can see (by finding the emulator run and downloading it's file and looking at it's results)

**Review by david-allison — APPROVED** [review]
Really clean implementation, thanks!

**Comment by voczi** [comment]
changes for jacoco.gradle, *slightly* overengineered but maybe better feedback in case of trouble?

```
// Issue 16640 - some emulators run, but register zero coverage
import groovy.xml.XmlParser
tasks.register('assertNonzeroAndroidTestCoverage') {
    doLast {
        File jacocoReport = layout.buildDirectory.dir("reports/jacoco/jacocoAndroidTestReport/jacocoAndroidTestReport.xml").get().asFile

        if (!jacocoReport.exists())
            throw new FileNotFoundException("jacocoAndroidTestReport.xml was not found after running jacocoAndroidTestReport")

        XmlParser xmlParser = new XmlParser()
        xmlParser.setFeature("http://apache.org/xml/features/disallow-doctype-decl", false)
        xmlParser.setFeature("http://apache.org/xml/features/nonvalidating/load-external-dtd", false)

        Node reportRoot = xmlParser.parse(jacocoReport)
        boolean hasCovered = false

        // https://github.com/jacoco/jacoco/blob/5aabb2eb60bbcd05df968005f1746ba19dcd5361/org.jacoco.report/src/org/jacoco/report/internal/xml/ReportElement.java#L190
        for (Node child : reportRoot.children()) {
            if (child.name() != "counter")
                continue;

            if (child.attribute("covered") == "0") {
                logger.warn("jacoco registered zero code coverage for counter type %s", child.attribute("type"))
            } else {
                hasCovered = true
            }
        }

        if (!hasCovered)
            throw new GradleScriptException("androidTest registered zero code coverage in jacocoAndroidTestReport.xml. Probably some incompatibilities in the toolchain.", null)
    }
}
```

File: AnkiDroid/jacoco.gradle [hunk 1]
```diff
@@ -167,3 +167,28 @@ def androidTestReport = tasks.register('jacocoAndroidTestReport', JacocoReport)
     ])
 }
 androidTestReport.configure { dependsOn('connectedPlayDebugAndroidTest') }
+
+// Issue 16640 - some emulators run, but register zero coverage
+tasks.register('assertNonzeroAndroidTestCoverage') {
```
Comment (voczi) [inline AnkiDroid/jacoco.gradle:172]: see comment above

Comment (mikehardy) [reply AnkiDroid/jacoco.gradle:172]: yeah - I like that better, didn't realize there was a nice built-in XML parser but that is certainly an improvement over pattern matching, will incorporate, thank you

**Comment by mikehardy** [comment]
re-pushed with two changes:

- used XML parser vs pattern matching, and tidier control-flow / error reporting from @voczi (thank you!)
- added a big screaming warning file in the copy target destination directory for the workaround so future developers will know where the files are coming from and how to work with this workaround, ideally avoiding surprise / wasted effort of any attempts to edit the copied files vs the source files

**Review by voczi — APPROVED** [review]

**Pull request merged**